### PR TITLE
Segy converter

### LINKED ISSUE #87: Support SEGY files that do not have geometry - consistent support of segy files
The current code uses segyio.tools.cube() to extract a volume of data from segy files that can easily be chuncked up. However, most segy files do not have geometry in them that allows segyio to infer the dimensions of the volume. In the case where geometry is not present. You need to manually infer the dimensions by reading through the trace headers and constructing a volume manually.

### PR DESCRIPTION
**Pull request by MJZawacki**
This PR addresses some issues when users bring their own SEG-Y seismic and label files as input to the training scripts. Most SEG-Y files do not have a uniform set of inline and crossline traces so segyio has problems inferring the geometry as it tries to generate a numpy array from the original traces.

## Details
* README
* Unit tests for the utilities
* Notebook to create test data and describe issues with segy files
* Notebook showing how to run the conversion script
* core logic in the utils.segyextract module
* cmd line script is convert_segy.py which can convert to single or multiple npy files. It also calculates stddev and mean of the original segy data that can be used for normalization and clipping

resolves #87

### CODE REVIEW

File: contrib/segyconverter/test/test_segyextract_eval_functions.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    # test
+    segyextract._identify_fast_direction(df, 'fast', 'slow')
+    # validate
+    assert(df.keys()[0] == 'fast')
+    assert(df.keys()[1] == 'slow')
+
+
+def test_identify_fast_direction_should_handle_xline_sequence_2():
+    # setup
+    df = pd.DataFrame({'i': [101, 102, 102, 102, 102, 102], 'j': [301, 301, 302, 303, 304, 305]})
```
Comment (maxkazmsft) [inline contrib/segyconverter/test/test_segyextract_eval_functions.py]: Same comment about reading from a common file

File: contrib/segyconverter/utils/normalize_cube.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+if __name__ == '__main__':
+    """
+        Normalize cube (numpy array) based on statistics. Normalized cube range will be within
+        [min, max] range
+
+        Sample call: python3 normalize_cube.py --cube=<cube> --stddev=<stddev_value> --max_range=<max_range_value>
+                    --min_range=<min_range_value> -k=<k>
+    """
+    parser = argparse.ArgumentParser(description='Normalize cube')
```
Comment (maxkazmsft) [inline contrib/segyconverter/utils/normalize_cube.py]: Do you want to try using Fire instead like we do in our scripts?

File: contrib/segyconverter/utils/segyextract.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        layer_slowmax = max(slow_distinct)
+        layer_slowmin = min(slow_distinct)
+        layer_trace_ids = trace_headers[(trace_headers.fast >= layer_fastmin)
+                                        & (trace_headers.fast <= layer_fastmax)
+                                        & (trace_headers.slow >= layer_slowmin)
+                                        & (trace_headers.slow <= layer_slowmax)]
+
+        block = np.zeros((len(fast_distinct), len(slow_distinct), sampledepth), dtype=np.float32)
+        for _, row in layer_trace_ids.iterrows():
+            block[(row[FAST] - layer_fastmin) // fast_line_space,
```
Comment (maxkazmsft) [inline contrib/segyconverter/utils/segyextract.py]: I see a lot of slow and fast in the code - can these be carried into a separate YACS config? so row[config.FAST] then names can be changed.

File: contrib/segyconverter/utils/segyextract.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        mean = np.mean(block)
+
+        with open(os.path.join(output_dir, prefix + '.json'), 'w') as f:
+            f.write(json.dumps({'stddev': str(stddev), 'mean': str(mean)}))
+        print("Npy files written: 1")
+    return block
+
+
+def process_segy_data(input_file, output_dir, prefix,
+                      iline=189, xline=193, n_points=128, stride=128):
```
Comment (maxkazmsft) [inline contrib/segyconverter/utils/segyextract.py]: similar comment about configs and hardcoding

File: contrib/segyconverter/utils/segyextract.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    :param int stride: stride when writing data
+    """
+    fast_indexes, slow_indexes, trace_headers, _ = get_segy_metadata(input_file, iline, xline)
+    with segyio.open(input_file, ignore_geometry=True) as segy_file:
+        segy_file.mmap()
+        # Global variance of segy data
+        variance = 0
+        mean = 0
+        sample_count = 0
+        filecount = 0
```
Comment (maxkazmsft) [inline contrib/segyconverter/utils/segyextract.py]: config again

File: contrib/segyconverter/segy_sample_files.ipynb [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
...srxr6+1zwYC2m2pw2+vU/R8kcBFwJIPTHBuB8ydbTv+S7A9cB5xdVU8Mb5uWeTDPGEzVPKiqbVW1DFgMvAJ40YRLGru5Y5DkxcC5DMbiHwHPBz40wRJ7leSNwKaqWj3pWiZlJ2MwNfOg86qqehnwegaXfbxmeOOz/blgQBvNw0kOBeh+bppwPWNXVQ93/1A/BVzM4MNqr5VkHwbB5Mqqur5rnqp5MN8YTNs82K6qHgNuAl4JHJhkYbdpMfDQxAobo6ExOLk7BV5VtQX4b+zd8+AE4E1JHgS+yODU5qeZrnnwjDFI8t+nbB5QVQ91PzcBX2bw9/b2uWBAG80NwOnd8unAVydYy0Rsn4CdNwN372jfPV13fcklwPqqumBo09TMgx2NwZTNg5kkB3bLzwNex+BavJuAt3a77e3zYL4xuGfoAykMrrnZa+dBVZ1bVYurailwKvDtqnonUzQPdjAG75qmeZBkvyQHbF8G/jmDv7e3z4WFu95luiS5CjgRODjJBuBjwCeBa5KcAfwYePvkKuzfDsbgxO4W6gIeBN43sQL7dwLwbuAH3bU3AB9huubBjsbgtCmaB4cClydZwOB/Zq+pqq8lWQd8McmfAXcyCLJ7qx2NwbeTzAAB1gBnTrLICfkQ0zMPduTKKZoHhwBfHmRRFgL/o6q+keT/0tPngk8SkCRJaoynOCVJkhpjQJMkSWqMAU2SJKkxBjRJkqTGGNAkSZIaY0CTtNdL8n9G2Oc7SWa75Ru3f/+XJE2C34Mmaa9XVX+wm/u/oa9aJGkUHkGTtNdL8vPu54ndkbJrk9yT5MruW9Dn7v9gkoOTLE2yPsnFSdYm+Wb3jfokOTLJN7oHJ/+vJFP3nE5J/TGgSZo2xwFnA8cARzB4asLOHAX8RVUdCzwG/KuufQXwb6vq5cC/Az7XT7mSppGnOCVNm9uragNA9xirpcCtO9n/R1W1/XFXq4GlSfYH/gD40tABuH37KVfSNDKgSZo2W4aWt7Hrfwfn7v88BmcfHquqZc9ybZIEeIpTknZbVT0B/CjJ2wAy8NIJlyVpL2JAk6TfzDuBM5J8H1gLnDLheiTtRVJVk65BkiRJQzyCJkmS1BgDmiRJUmMMaJIkSY0xoEmSJDXGgCZJktQYA5okSVJjDGiSJEmNMaBJkiQ15v8DYqF0avnwe1gAAAAASUVORK5CYII=\n",
```
Comment (maxkazmsft) [inline contrib/segyconverter/segy_sample_files.ipynb]: I'll review this later again - dep on image size we may want to host on Azure.

File: contrib/segyconverter/test/test_convert_segy.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+def test_convert_segy_generates_single_npy(tmpdir):
+    # Setup
+    prefix = 'volume1'
+    input_file = TESTFILE
+    output_dir = tmpdir.strpath
+    metadata_only = False
+    iline = 189
+    xline = 193
+    cube_size = -1
+    stride = 128
```
Comment (maxkazmsft) [inline contrib/segyconverter/test/test_convert_segy.py]: config again

File: contrib/segyconverter/test/test_convert_segy.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+import os
+import numpy as np
+import convert_segy
+import test_util
+import pytest
+
+MAX_RANGE = 1
+MIN_RANGE = 0
+ERROR_EXIT_CODE = 99
+TESTFILE = "./contrib/segyconverter/test/test_data/volume1-label.segy"
```
Comment (maxkazmsft) [inline contrib/segyconverter/test/test_convert_segy.py]: Config

File: contrib/segyconverter/segy_sample_files.ipynb (hunk 6 above)
Comment (MJZawacki) [reply contrib/segyconverter/segy_sample_files.ipynb]: This is the output from the scatter plot of the segy files so I don't think there is a way to host on azure storage. Do you want the notebooks checked in with output data? or should the user be expected  to run the notebook to see the output?

File: contrib/segyconverter/utils/segyextract.py (hunk 5 above)
Comment (MJZawacki) [reply contrib/segyconverter/utils/segyextract.py]: These variables are not configurable. They are initializing counters to calculate the standard deviation if the file is too big to load into memory

File: contrib/segyconverter/segy_sample_files.ipynb (hunk 6 above)
Comment (MJZawacki) [reply contrib/segyconverter/segy_sample_files.ipynb]: The notebook with these images is only 60K so I think it shouldn't add too much bloat

**Comment by MJZawacki** [comment]
Replaced by #192 w/ no binary files

**Pull request closed without merging**